### ISSUE TITLE
OPL3 and 486 related changes

### ISSUE THREAD

**Issue by Doridian** [jdah/tetris-os#26]
Given this is an old-school game, it clearly should run on a 486.
So I made some slight adjustments to the code to optimize as much as possible given the limited hardware.
Also changed out the SB16 code for AdLib/OPL3 code (still working on the parameters to fine-tune it), because there is no way I can make the poor 486 generated waveforms fast enough for 48kHz PCM.

See my fork here: https://github.com/Doridian/tetris-os

If there is any interest to port parts of this code into this repo, let me know, I can split the commits up (I doubt you want all of the changes, especially the complete change in sound API)

**Comment by daanmsvl** [comment]
How did you get the ISO working? I made an ISO following the instructions, used imgburn to burn it to a CD and put it into my Pentium-133 which can boot from CD but it doesn't recognize it as a bootable CD. I was anxious to try it since this machine has a SB Pro II

**Comment by Doridian** [comment]
> How did you get the ISO working? I made an ISO following the instructions, used imgburn to burn it to a CD and put it into my Pentium-133 which can boot from CD but it doesn't recognize it as a bootable CD. I was anxious to try it since this machine has a SB Pro II

Since the MiSTer is an FPGA simulated machine, I in fact did not. I renamed the file to .vhd and mounted it that way, that worked.
I could not get it to recognize the bootsector if I mounted it as a CD-ROM.
Floppy also worked but sadly floppy doesn't support INT 13h 42h (extended read), so I could actually boot from it beyond "DISK ERROR".
Your best bet would probably be to see if you have a CF-IDE adapter to just smash the image raw onto an SD card.
I have the suspicion that booting from CD-ROM needs some special sauce in the ISO beyond an MBR marker

**Comment by jdah** [comment]
Wow you've really been up in the code! I would definitely appreciate integrating some of these changes into the codebase - probably not everything to go along with the alternative sound system/lack of FPU, but some of these speed improvements would be very welcome.

I'll see about switching away from the extended read interrupt towards something floppy-compatible.

I also think some quick speed gains could be made to the music system if we just halved the frequency to a 24 khz sample rate or maybe even lower.

**Comment by Doridian** [comment]
> Wow you've really been up in the code! I would definitely appreciate integrating some of these changes into the codebase - probably not everything to go along with the alternative sound system/lack of FPU, but some of these speed improvements would be very welcome.
> 
> I'll see about switching away from the extended read interrupt towards something floppy-compatible.
> 
> I also think some quick speed gains could be made to the music system if we just halved the frequency to a 24 khz sample rate or maybe even lower.

Well, the good news is...I no longer even need softfp, as all it was used for was PCM, with the AdLib the soundcard does the work, soooo. We could probably make it switchable between SB16 and OPL3, one requiring FPU, the other not with some ifdefs.
If you want to do that, feel free, I'm still poking at
- Getting an actual CDROM ISO that boots on real machines (el torito headers)
- Better synth settings for the OPL3 music. With all the FM modulations this can sound amazing

**Comment by jdah** [comment]
> We could probably make it switchable between SB16 and OPL3, one requiring FPU, the other not with some ifdefs.

That could be easy enough with a stable API on `sound.c` that both the SB16 and OPL3 drivers could implement.

**Comment by Doridian** [comment]
> > We could probably make it switchable between SB16 and OPL3, one requiring FPU, the other not with some ifdefs.
> 
> That could be easy enough with a stable API on `sound.c` that both the SB16 and OPL3 drivers could implement.

Sounds great.
I just got done finding an okay setting (electric piano + "gun shot" for the percussion, sounds pretty alright, could be better, but I'm not a musician...)

**Comment by zment4** [comment]
> > We could probably make it switchable between SB16 and OPL3, one requiring FPU, the other not with some ifdefs.
> 
> That could be easy enough with a stable API on `sound.c` that both the SB16 and OPL3 drivers could implement.

Yes please to a stable sound API, I can then add at least a simple pc speaker if not a more complex and perhaps I could look into AC97 as well, depending on how complicated it gets and how motivated I am 😁

**Comment by Doridian** [comment]
> > > We could probably make it switchable between SB16 and OPL3, one requiring FPU, the other not with some ifdefs.
> > 
> > 
> > That could be easy enough with a stable API on `sound.c` that both the SB16 and OPL3 drivers could implement.
> 
> Yes please to a stable sound API, I can then add at least a simple pc speaker if not a more complex and perhaps I could look into AC97 as well, depending on how complicated it gets and how motivated I am 😁

My PR...which has turned into a bit of a monster now, since I was lazy and didn't nicely seperate changes...
Does have a "sound API" (as you can see, it just uses the same function names for SB16 and OPL3, basically), so you can #define in the one you want to be enabled.
Probably could somehow PCSpeaker it, but....don't know how hard that'd be.
